km/km_init_guest.c

```diff
@@ -321,10 +319,14 @@ int km_clone(km_vcpu_t* vcpu,
             new_vcpu->vcpu_id,
             new_vcpu->regs.rip,
             new_vcpu->regs.rsp);
+
+   km_lock_vcpu_thr(new_vcpu);
    if (km_run_vcpu_thread(new_vcpu) < 0) {
+      km_unlock_vcpu_thr(new_vcpu);
       km_vcpu_put(new_vcpu);
```

Comment:
Why is it ok to not hold:

km_mutex_lock(&machine.vm_vcpu_mtx);

when you call km_vcpu_put() here?
It is held when km_vcpu_put() is called elsewhere.

Reply:
This is cleanup in case of failure to start `new_vcpu` thread. Nobody is going to look at it because of its state. I guess we can take the lock, it won't hurt. Same in `km_vcpu_fini()` - there is only one thread by then. I can move the mutex manipulation back inside the function I guess.

Reply:
If we have 2 or more threads creating threads concurrently it seems like we could run into situations where they are both doing things with the machine.vm_idle_vcpus.head concurrently, especially if thread creation is failing for some global reason affecting all thread creators.

If something that bad is happening, this little serialization faux pas is kind of irrelevant, the entire system will be a pile of dodo shortly.